app/uk/gov/hmrc/apisubscriptionfields/model/Model.scala

```diff
@@ -28,35 +28,43 @@ case class ApiVersion(value: String) extends AnyVal
 
 case class SubscriptionFieldsId(value: UUID) extends AnyVal
 
-object SubscriptionIdentifier {
-  private val Separator = "##"
-  private type SeparatorType = String
-
-  def decode(text:String): Option[SubscriptionIdentifier] = {
-    def findSeparator(separatorToFind: SeparatorType) : SeparatorType = {
-      if (text.split(separatorToFind).length > 3)
-        findSeparator(separatorToFind+Separator)
-      else
-        separatorToFind
-    }
-
-    val parts = text.split(findSeparator(Separator))
-    Some(SubscriptionIdentifier(AppId(parts(0)), ApiContext(parts(1)), ApiVersion(parts(2))))
-  }
+object SubscriptionIdentifier extends Decoder[SubscriptionIdentifier] {
+  override val separator = Separator
+  override val numOfParts: Int = 3
+
+  override protected def decode(tokens: Seq[String]): SubscriptionIdentifier =
+    SubscriptionIdentifier(AppId(tokens(0)), ApiContext(tokens(1)), ApiVersion(tokens(2)))
+}
+
+case class SubscriptionIdentifier(applicationId: AppId, apiContext: ApiContext, apiVersion: ApiVersion) extends Encoder[SubscriptionIdentifier] {
+  override val separator = Separator
+
+  def encode(): String =
+    encode(applicationId.value, apiContext.value, apiVersion.value)
+
+}
+
+object FieldsDefinitionIdentifier extends Decoder[FieldsDefinitionIdentifier] {
+  override val separator = Separator
+  override val numOfParts: Int = 2
+
+  override protected def decode(tokens: Seq[String]): FieldsDefinitionIdentifier =
+    FieldsDefinitionIdentifier(ApiContext(tokens(0)), ApiVersion(tokens(1)))
+}
+
+case class FieldsDefinitionIdentifier(apiContext: ApiContext, apiVersion: ApiVersion) extends Encoder[FieldsDefinitionIdentifier] {
+  override val separator = Separator
+
+  def encode(): String = encode(apiContext.value, apiVersion.value)
 }
 
-case class SubscriptionIdentifier(applicationId: AppId, apiContext: ApiContext, apiVersion: ApiVersion) {
-  import SubscriptionIdentifier._
+object FieldDefinitionType extends Enumeration {
+  type FieldDefinitionType = Value
 
-  def encode(): String = {
-    @tailrec
-    def findSeparator(text: String)(separatorToFind: SeparatorType): SeparatorType =
-      if (!text.contains(separatorToFind))
-        separatorToFind
-      else
-        findSeparator(text)(separatorToFind+Separator)
+  val URL = Value("URL")
```

Comment:
We might need to review this way to define the type of the fields definitions.
But let's keep this review for later.